src/client/Microsoft.Identity.Client/Internal/Requests/CredentialBasedMsiAuthRequest.cs

```diff
@@ -0,0 +1,329 @@
+            catch (MsalServiceException ex)
+            {
+                logger.Verbose(() => $"[CredentialBasedMsiAuthRequest] Caught an exception. {ex.Message}. Error Code : {ex.ErrorCode} Status Code : {ex.StatusCode}");
+
+                exception = MsalServiceExceptionFactory.CreateManagedIdentityException(
+                    ex.ErrorCode,
+                    ex.Message,
+                    ex.InnerException,
+                    ManagedIdentitySource.Credential,
+                    ex.StatusCode);
+
+                throw exception;
+            }
+            catch (Exception e) when (e is not MsalServiceException)
+            {
+                logger.Error($"[Managed Identity] Exception: {e.Message}");
+                exception = e;
+                message = MsalErrorMessage.CredentialEndpointNoResponseReceived;
+            }
+
+            MsalException msalException = MsalServiceExceptionFactory.CreateManagedIdentityException(
+                MsalError.CredentialRequestFailed,
+                message,
+                exception,
+                ManagedIdentitySource.Credential,
+                null);
+
+            throw msalException;
+        }
+
+        private async Task<CredentialResponse> GetCredentialAssertionAsync(
+            IKeyMaterialManager keyMaterial,
+            ILoggerAdapter logger,
+            CancellationToken cancellationToken
+            )
+        {
+            var managedIdentityCredentialResponse = new ManagedIdentityCredentialResponse(
+                _credentialEndpoint,
+                keyMaterial.BindingCertificate,
+                AuthenticationRequestParameters.RequestContext,
+                cancellationToken);
+
+            CredentialResponse credentialResponse = await managedIdentityCredentialResponse.GetCredentialAsync().ConfigureAwait(false);
+
+            logger.Verbose(() => "[CredentialBasedMsiAuthRequest] A credential was successfully fetched.");
+
+            return credentialResponse;
+        }
+
+        /// <summary>
+        /// Creates an OAuth2 client request for fetching the managed identity credential.
+        /// </summary>
+        /// <param name="keyMaterial"></param>
+        /// <param name="httpManager"></param>
+        /// <param name="credentialResponse"></param>
+        /// <returns></returns>
+        private OAuth2Client CreateClientRequest(
+            IKeyMaterialManager keyMaterial,
+            IHttpManager httpManager,
+            CredentialResponse credentialResponse)
+        {
+            // Initialize an OAuth2 client with logger, HTTP manager, and binding certificate.
+            var client = new OAuth2Client(
+                AuthenticationRequestParameters.RequestContext.Logger,
+                httpManager,
+                keyMaterial.BindingCertificate);
+
+            // Convert overridden scopes to a single string.
+            string scopes = GetOverriddenScopes(AuthenticationRequestParameters.Scope).AsSingleString();
```

Comment:
Can there be multiple scopes? From the API we will be accepting a single scope / resource. Are we expecting any scopes back from credential endpoint?

Reply:
this will be same like the cca flow, only one scope